doc/IPset.md

```diff
@@ -2,218 +2,709 @@ DPVS IPset
+
+```bash
+# ./bin/dpip ipset add bar 2001::1,8080-8082,2002::aaaa:bbbb:ccc0:0/108 nomatch
+# ./bin/dpip ipset -v list bar
+Name: bar
+Type: hash:ip,port,net
+Header: family inet6  hashsize 1024  maxelem 65535  
+Size in memory: 17296
+References: 0
+Number of entries: 6
+Members:
+2001::1,unspec:8080,2002::aaaa:bbbb:ccc0:0/108 nomatch  
+2001::1,unspec:8081,2002::aaaa:bbbb:ccc0:0/108 nomatch  
+2001::1,unspec:8082,2002::aaaa:bbbb:ccc0:0/108 nomatch  
+2001::1,unspec:8080,2002::/64  
+2001::1,unspec:8081,2002::/64  
+2001::1,unspec:8082,2002::/64 
+# ./bin/dpip ipset test bar 2001::1,8081,2002::1:2:3:4
+true
+# ./bin/dpip ipset test bar 2001::1,8081,2002::1:2:3:4:5
+false
+# ./bin/dpip ipset test bar 2001::1,8081,2002::aaaa:bbbb:ccc1:2345 
+false
+```
+
+<a id='hash_net_port_net'/>
+
+#### hash:net,port,net
+
+The hash:net,port,net set type uses a hash table to store two different sized IP network addresses and a port number triples. It is similar to the hash:ip,port,net set type, except that the first segment indicates a network address rather than a single IP address. Option "nomatch" also supported, but should be used with caution. The first "net" segment takes precedence over the second one when matching against the set, regardless of whether the entry is of "nomatch" option.
+
+```bash
+# ./bin/dpip ipset create foo hash:net,port,net
+# ./bin/dpip ipset add foo 10.60.0.0/16,tcp:10240-10242,10.130.0.0/16
+# ./bin/dpip ipset add foo 10.60.100.100-10.60.100.200,tcp:10242,10.130.100.0/24 nomatch
+# ./bin/dpip -v ipset show foo
+Name: foo
+Type: hash:net,port,net
+Header: family inet  hashsize 1024  maxelem 65535  
+Size in memory: 17752
+References: 0
+Number of entries: 9
+Members:
+10.60.100.200/32,tcp:10242,10.130.100.0/24 nomatch  
+10.60.100.100/30,tcp:10242,10.130.100.0/24 nomatch  
+10.60.100.104/29,tcp:10242,10.130.100.0/24 nomatch  
+10.60.100.192/29,tcp:10242,10.130.100.0/24 nomatch  
+10.60.100.112/28,tcp:10242,10.130.100.0/24 nomatch  
+10.60.100.128/26,tcp:10242,10.130.100.0/24 nomatch  
+10.60.0.0/16,tcp:10240,10.130.0.0/16  
+10.60.0.0/16,tcp:10241,10.130.0.0/16  
+10.60.0.0/16,tcp:10242,10.130.0.0/16
+# ./bin/dpip ipset test foo 10.60.0.0,tcp:10242,10.130.255.255
+true
+# ./bin/dpip ipset test foo 10.60.100.111,tcp:10240,10.130.3.4
+true
+# ./bin/dpip ipset test foo 10.60.100.111,tcp:10240,10.130.100.222
```

Comment:
Is there an example interpreting "The first "net" segment takes precedence over the second one when matching against the set, regardless of whether the entry is of "nomatch" option"?

Reply:
add an example:

```
# ./bin/dpip ipset flush foo 
# ./bin/dpip ipset add foo 192.168.100.0/24,udp:6000,192.168.200.0/25
# ./bin/dpip ipset add foo 192.168.100.0/25,udp:6000,192.168.200.0/24 nomatch
# ./bin/dpip ipset list foo 
Name: foo 
Type: hash:net,port,net
Header: family inet  hashsize 1024  maxelem 65535  
Size in memory: 16688
References: 0
Number of entries: 2
Members:
192.168.100.0/24,udp:6000,192.168.200.0/25  
192.168.100.0/25,udp:6000,192.168.200.0/24 nomatch  
# ./bin/dpip ipset test foo 192.168.100.1,udp:6000,192.168.200.1
false
# ./bin/dpip ipset flush foo
# ./bin/dpip ipset add foo 192.168.100.0/24,udp:6000,192.168.200.0/25 nomatch
# ./bin/dpip ipset add foo 192.168.100.0/25,udp:6000,192.168.200.0/24
# ./bin/dpip ipset list foo
Name: foo
Type: hash:net,port,net
Header: family inet  hashsize 1024  maxelem 65535
Size in memory: 16688
References: 0
Number of entries: 2
Members:
192.168.100.0/24,udp:6000,192.168.200.0/25 nomatch
192.168.100.0/25,udp:6000,192.168.200.0/24
[root@10.15.204.12 dpvs]# ./bin/dpip ipset test foo 192.168.100.1,udp:6000,192.168.200.1
true
```